pokedex/simulator/difftest/src/replay.rs

```diff
@@ -0,0 +1,213 @@
+            writeln!(f, "[{id}]({name})={val}")?;
+        }
+        writeln!(f, "{}", "-".repeat(80))?;
+
+        Ok(())
+    }
+}
+
+impl CpuState {
+    /// Return an unintiallze CpuState. Reset and emulator alignment should be handled on software side.
+    pub fn new() -> Self {
+        Self {
+            gpr: [0; 32],
+            fpr: [0; 32],
+            vregs: Vec::new(),
+            pc: 0,
+            csr: HashMap::new(),
+            is_poweroff: false,
+        }
+    }
+
+    /// Update shadow GPR, return old data if the written data is different with it.
+    pub(crate) fn write_gpr(&mut self, rd: usize, val: u32) -> Option<u32> {
+        assert!(rd > 0 && rd < 32);
+        let old_val = self.gpr[rd];
+        if old_val != val {
+            self.gpr[rd] = val;
+            Some(old_val)
+        } else {
+            None
+        }
+    }
+
+    /// Update shadow FPR, return old data if the written data is different with it.
+    pub(crate) fn write_fpr(&mut self, rd: usize, val: u32) -> Option<u32> {
+        assert!(rd < 32);
+        let old_val = self.fpr[rd];
+        if old_val != val {
+            self.fpr[rd] = val;
+            Some(old_val)
+        } else {
+            None
+        }
+    }
+
+    pub(crate) fn write_csr(&mut self, name: &str, id: u16, val: u32) -> Option<u32> {
+        let (_, entry) = self.csr.entry(id).or_insert((name.to_string(), val));
+        let old_val = *entry;
+        if old_val != val {
+            *entry = val;
+            Some(old_val)
+        } else {
+            None
+        }
+    }
+
+    pub(crate) fn write_vreg(&mut self, data: &[u8], mask: &[bool]) -> Option<Vec<u8>> {
+        todo!()
+    }
+}
+
+/// CSR Write is conservative, it can be an implicit write from instruction or an explicit write
+/// request by Zicsr instructions. Distinguish those writes could speed up diff test.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum CsrCheckType {
+    NoWrite,
+    AllCsr,
+    FpCsrOnly,
+    VecCsrOnly,
+}
+
+impl Default for CsrCheckType {
+    fn default() -> Self {
+        Self::NoWrite
+    }
+}
+
+impl CsrCheckType {
+    pub(crate) fn has_write(&self) -> bool {
+        !(matches!(self, Self::NoWrite))
+    }
+}
+
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
+pub struct StateCheckType {
+    // rd for general propose register
+    pub(crate) gpr_rd: Option<usize>,
+    // rd for floating point register
+    pub(crate) fpr_rd: Option<usize>,
+    // written data index mask for the concatenated vector register
+    pub(crate) vreg_mask: Option<u32>,
+    // possible write type for CSR
+    pub(crate) csr_mask: CsrCheckType,
+}
+
+pub struct CommitCassette<'a, 'b, T>
+where
+    T: IsCpuState,
+{
+    commit_cursor: &'a mut Peekable<Iter<'b, T>>,
+    state: CpuState,
+}
+
+pub trait IsCpuState {
+    fn get_pc(&self) -> u32;
+    fn has_reset(&self) -> Option<u32>;
+    fn write_cpu_state(&self, state: &mut CpuState) -> StateCheckType;
+}
```

Comment:
Maybe it should be called "IsCommit". I see it's impl-ed in Commit and InsnCommit, not CpuState.